packages/reporters/dev-server/src/Server.js

```diff
@@ -0,0 +1,204 @@
+// @flow
+import type {Request, Response, DevServerOptions} from './types.js.flow';
+import type {BundleGraph} from '@parcel/types';
+import type {PrintableError} from '@parcel/reporter-cli/src/prettyError';
+
+import EventEmitter from 'events';
+import path from 'path';
+import http from 'http';
+import https from 'https';
+import url from 'url';
+import serveStatic from 'serve-static';
+import getPort from 'get-port';
+import ansiHtml from 'ansi-html';
+import logger from '@parcel/logger';
+import prettyError from '@parcel/reporter-cli/src/prettyError';
+import generateCertificate from '@parcel/server-utils/src/generateCertificate';
+import getCertificate from '@parcel/server-utils/src/getCertificate';
+import serverErrors from './serverErrors';
+import {readFile} from '@parcel/fs';
+
+function setHeaders(res: Response) {
+  res.setHeader('Access-Control-Allow-Origin', '*');
+  res.setHeader(
+    'Access-Control-Allow-Methods',
+    'GET, HEAD, PUT, PATCH, POST, DELETE'
+  );
+  res.setHeader(
+    'Access-Control-Allow-Headers',
+    'Origin, X-Requested-With, Content-Type, Accept, Content-Type'
+  );
+}
+
+export default class Server extends EventEmitter {
+  pending: boolean;
+  options: DevServerOptions;
+  serve: (
+    req: Request,
+    res: Response,
+    next?: (req: Request, res: Response, next?: (any) => any) => any
+  ) => any;
+  bundleGraph: BundleGraph | null;
+  error: PrintableError | null;
+
+  constructor(options: DevServerOptions) {
+    super();
+
+    this.options = options;
+    this.pending = true;
+    this.bundleGraph = null;
+    this.error = null;
+
+    this.serve = serveStatic(this.options.distDir, {
+      index: false,
+      redirect: false,
+      setHeaders: setHeaders,
+      dotfiles: 'allow'
+    });
+  }
+
+  buildSuccess(bundleGraph: BundleGraph) {
+    this.bundleGraph = bundleGraph;
+    this.pending = false;
+
+    this.emit('bundled');
+  }
+
+  buildError(error: PrintableError) {
+    this.error = error;
+  }
+
+  respond(req: Request, res: Response) {
+    let {pathname} = url.parse(req.url);
+
+    if (this.error) {
+      return this.send500(req, res);
+    } else if (
+      !pathname ||
+      !pathname.startsWith(this.options.publicUrl) ||
+      path.extname(pathname) === ''
+    ) {
+      // If the URL doesn't start with the public path, or the URL doesn't
+      // have a file extension, send the main HTML bundle.
+      return this.sendIndex(req, res);
+    } else {
+      // Otherwise, serve the file from the dist folder
+      req.url = pathname.slice(this.options.publicUrl.length);
+      return this.serve(req, res, () => this.sendIndex(req, res));
+    }
+  }
+
+  async sendIndex(req: Request, res: Response) {
+    if (this.bundleGraph) {
+      // If the main asset is an HTML file, serve it
+      let htmlBundle = null;
+      this.bundleGraph.traverseBundles(bundle => {
+        if (bundle.type === 'html' && bundle.isEntry) {
+          htmlBundle = bundle;
```

Comment:
The third argument to the traversal function is a traversal context object which has a `stop` method. Should be slightly faster to stop once you’ve found the bundle you need.

Reply:
Also, what should we do about multiple entry points here? We don’t currently handle that case in parcel 1 but if a user has multiple html files as entries, we should pick the one named `index.html` by default. If there is no index.html, then take the first entry point. There is an issue for that somewhere...

Reply:
> There is an issue for that somewhere...

https://github.com/parcel-bundler/parcel/issues/1315